test/model/internal_link_test.dart

```diff
@@ -0,0 +1,394 @@
+      checkExpectedStreamNarrow('#narrow/stream/1nonsense-general', null);
+      checkExpectedStreamNarrow('#narrow/stream/-general',          null);
+    });
+  });
+
+  group('ParseInternalLink with historic links', () {
+    group('for stream with hyphens or even looking like new-style', () {
+      final streams = [
+        eg.stream(streamId: 1, name: 'test-team'),
+        eg.stream(streamId: 2, name: '311'),
+        eg.stream(streamId: 3, name: '311-'),
+        eg.stream(streamId: 4, name: '311-help'),
+        eg.stream(streamId: 5, name: '--help'),
+      ];
+      const testCases = [
+        ('#narrow/stream/test-team/', StreamNarrow(1)),
+        ('#narrow/stream/311/',       StreamNarrow(2)),
+        ('#narrow/stream/311-/',      StreamNarrow(3)),
+        ('#narrow/stream/311-help/',  StreamNarrow(4)),
+        ('#narrow/stream/--help/',    StreamNarrow(5)),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('on ambiguous new- or old-style: new wins', () {
+      final streams = [
+        eg.stream(streamId: 1,   name: '311'),
+        eg.stream(streamId: 2,   name: '311-'),
+        eg.stream(streamId: 3,   name: '311-help'),
+        eg.stream(streamId: 311, name: 'collider'),
+      ];
+      const testCases = [
+        ('#narrow/stream/311/',      StreamNarrow(311)),
+        ('#narrow/stream/311-/',     StreamNarrow(311)),
+        ('#narrow/stream/311-help/', StreamNarrow(311)),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('on old stream link', () {
+      final streams = [
+        eg.stream(streamId: 1, name: 'check'),
+        eg.stream(streamId: 2, name: 'bot testing'),
+        eg.stream(streamId: 3, name: 'check.API'),
+        eg.stream(streamId: 4, name: 'stream'),
+        eg.stream(streamId: 5, name: 'topic'),
+      ];
+      const testCases = [
+        ('#narrow/stream/check/',         StreamNarrow(1)),
+        ('#narrow/stream/bot.20testing/', StreamNarrow(2)),
+        ('#narrow/stream/check.2EAPI/',   StreamNarrow(3)),
+        ('#narrow/stream/stream/',        StreamNarrow(4)),
+        ('#narrow/stream/topic/',         StreamNarrow(5)),
+
+        ('#narrow/stream/check.API/',     null),
+      ];
+      checkExpectedNarrows(testCases, streams: streams);
+    });
+
+    group('on old stream link, without realm info', () {
+      final stream = eg.stream(name: 'example');
+      final testCases = [
+        ('/#narrow/stream/${stream.name}/', StreamNarrow(stream.streamId)),
+        ('#narrow/stream/${stream.name}/',  StreamNarrow(stream.streamId)),
+      ];
+      checkExpectedNarrows(testCases, streams: [stream]);
+    });
+  });
+
+  group('ParseInternalLink', () {
+    group('topic link parsing', () {
+      final stream = eg.stream(name: "example");
+
+      group('basic', () {
+        String mkUrl(operand) {
+          return '#narrow/stream/${stream.streamId}-general/topic/$operand';
```

Comment:
This is a bit incongruous: the stream's name is "example" but the URL says "general". Typically the URL would accurately reflect the stream name. Our tests should generally use representative data except when the test is specifically about some way of being non-representative, so these should agree.

(And other tests above cover the possibility that the URL doesn't reflect the stream name.)